src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -91,117 +93,190 @@ export class KernelExecution implements IDisposable {
-                            // Cancel and remove from map pending cells.
-                            return this.cancelCell(cell.cell);
-                        })
-                    );
-                    break;
-                }
-            }
+            await this.pendingExecution.chainExecution(async () => this.executeQueuedCells(notebookPromise, document));
         } finally {
-            traceInfo(`Cancel all remaining cells after finally`);
-            await Promise.all(
-                codeCellsToExecute.map((cell) => {
-                    // Cancel and remove from map pending cells.
-                    return this.cancelCell(cell.cell);
-                })
-            );
             this.documentExecutions.delete(document);
             traceInfo('Restore notebook state to idle');
             await editor.edit((edit) =>
                 edit.replaceMetadata({ ...document.metadata, runState: vscodeNotebookEnums.NotebookRunState.Idle })
             );
         }
     }
-
-    public async cancelCell(cell: NotebookCell) {
-        const execution = this.cellExecutions.get(cell);
-        if (execution) {
-            this.cellExecutions.delete(cell);
-            traceCellMessage(cell, 'Cancel cell from Kernel Execution');
-            await execution.cancel();
-        } else {
-            traceCellMessage(cell, 'Cannot cancel cell execution from Kernel Execution');
+    public async interrupt(notebookPromise: Promise<INotebook>, document: NotebookDocument): Promise<InterruptResult> {
+        const notebook = await notebookPromise;
+        if (!notebook) {
+            traceInfo('No notebook to interrupt');
+            return InterruptResult.Success;
         }
+        traceInfo('Interrupt kernel execution');
+        // Interrupt all the cells, & wait for all to complete.
+        const cellExecutionPromises = this.waitForAllQueuedCells(document, true);
+
+        // Interrupt the active execution
+        const result = this._interruptPromise
+            ? await this._interruptPromise
+            : await (this._interruptPromise = this.interruptExecution(notebook.session, cellExecutionPromises));
+
+        // Done interrupting, clear interrupt promise
+        this._interruptPromise = undefined;
+
+        return result;
+    }
+    public dispose() {
+        this.disposables.forEach((d) => d.dispose());
+    }
+    private async cancelAllCells(notebookPromise: Promise<INotebook>, document: NotebookDocument): Promise<void> {
+        await this.interrupt(notebookPromise, document);
+        await this.waitForAllQueuedCells(document);
     }
 
-    public async cancelAllCells(document: NotebookDocument): Promise<void> {
-        if (this.documentExecutions.get(document)) {
-            this.documentExecutions.get(document)!.cancel();
+    private async waitForAllQueuedCells(document: NotebookDocument, interrupt = false) {
+        traceInfo('Interrupt kernel execution');
```

Comment:
This traceInfo is confusing. I'm not entirely sure what this method does? REturns the list of promises for executing cells?

Reply:
Fixed, renamed method.